contracts/libraries/VaultLifecycleSTETH.sol

```diff
@@ -0,0 +1,747 @@
+        // transfers the rest in `asset`
+        if (withdrawAmount > yieldTokenBalance) {
+            withdrawBaseToken(
+                weth,
+                asset,
+                collateralToken,
+                recipient,
+                withdrawAmount,
+                yieldTokenBalance
+            );
+        }
+    }
+
+    /**
+     * @notice Withdraws yvWETH from vault
+     * @param collateralToken is the address of the collateral token
+     * @param recipient is the recipient
+     * @param withdrawAmount is the withdraw amount in terms of yearn tokens
+     */
+    function withdrawYieldToken(
+        address collateralToken,
+        address recipient,
+        uint256 withdrawAmount
+    ) internal returns (uint256 yieldTokenBalance) {
+        IERC20 collateral = IERC20(collateralToken);
+
+        yieldTokenBalance = collateral.balanceOf(address(this));
+        uint256 yieldTokensToWithdraw =
+            dsmin(yieldTokenBalance, withdrawAmount);
+        if (yieldTokensToWithdraw > 0) {
+            collateral.safeTransfer(recipient, yieldTokensToWithdraw);
+        }
+    }
+
+    /**
+     * @notice Withdraws `asset` from vault
+     * @param weth is the weth address
+     * @param asset is the vault asset address
+     * @param collateralToken is the address of the collateral token
+     * @param recipient is the recipient
+     * @param withdrawAmount is the withdraw amount in terms of yearn tokens
+     * @param yieldTokenBalance is the collateral token (yvWETH) balance of the vault
+     */
+    function withdrawBaseToken(
+        address weth,
+        address asset,
+        address collateralToken,
+        address recipient,
+        uint256 withdrawAmount,
+        uint256 yieldTokenBalance
+    ) internal {
+        uint256 underlyingTokensToWithdraw =
+            IWSTETH(collateralToken).getStETHByWstETH(
+                withdrawAmount.sub(yieldTokenBalance)
+            );
+
+        transferAsset(
+            weth,
+            asset,
+            payable(recipient),
+            underlyingTokensToWithdraw
+        );
+    }
+
+    /**
+     * @notice Unwraps the necessary amount of the yield-bearing yearn token
+     *         and transfers amount to vault
+     * @param amount is the amount of `asset` to withdraw
+     * @param asset is the vault asset address
+     * @param collateralToken is the address of the collateral token
+     */
+    function unwrapYieldToken(
+        uint256 amount,
+        address asset,
+        address collateralToken
+    ) external {
+        uint256 assetBalance = IERC20(asset).balanceOf(address(this));
+
+        uint256 amountToUnwrap =
+            IWSTETH(collateralToken).getWstETHByStETH(
+                dsmax(assetBalance, amount).sub(assetBalance)
+            );
+
+        if (amountToUnwrap > 0) {
+            IWSTETH wsteth = IWSTETH(collateralToken);
+            // Unrap to stETH
+            wsteth.unwrap(wsteth.balanceOf(address(this)));
```

Comment:
you mentioned high gas bc we are swapping. we could potentially swap more than necessary (like 1.25x) so that we have more eth on hand. 

do we prioritize maximal efficiency or gas fees for user withdrawals.

Reply:
I'm fine with how we're doing it now, as long we don't go over 200k gas it should be fine